flow/project.py

```diff
@@ -2498,32 +2763,112 @@ def _guess_cmd(func, name, **kwargs):
+            groups.extend(getattr(cls, '_GROUPS', []))
+
+        # Initialize all groups without operations
+        for group in groups:
+            # Raise error for duplicate groups
+            if group['name'] in self._groups:
+                raise ValueError(
+                    "Repeat definition of group with name " +
+                    "'{}'.".format(group['name']))
+
+            self._groups[group['name']] = FlowGroup(path=self._path, **group)
+
+        # Add operations to group
+        # Currently recreating operation list. There may be a better way to do
+        # this.
+        operations = []
+        for cls in type(self).__mro__:
+            operations.extend(getattr(cls, '_OPERATION_FUNCTIONS', []))
+        for (name, func) in operations:
+            if hasattr(func, '_flow_group'):
+                for group in func._flow_group:
+                    self._groups[group].add_operation(name,
+                                                      self._operations[name])
+
+            # For singleton groups add directives
+            try:
+                self._groups[name].directives = func._flow_directives
+            except (AttributeError):
+                pass
+
+    def _get_operation_path(self, op_name):
+        operations = []
+        for cls in type(self).__mro__:
+            operations.extend(getattr(cls, '_OPERATION_FUNCTIONS', []))
+        func = [f for n, f in operations if n == op_name]
+        if func == list():
+            raise ValueError("{} not found in OPERATION_FUNCTIONS : "
+                             "{}".format(op_name,
+                                         ' '.join([name for name, _ in
+                                                   self._OPERATION_FUNCTIONS]))
+                             )
+        return getattr(func[0], '_flow_path',
+                       inspect.getsourcefile(inspect.getmodule(func[0])))
+
     @property
     def operations(self):
         "The dictionary of operations that have been added to the workflow."
         return self._operations
 
-    def _eligible_for_submission(self, job_operation):
-        """Determine if a job-operation is eligible for submission.
+    @property
+    def groups(self):
+        return self._groups
+
+    def eligible_for_submission(self, flow_group, job):
+        """Determine if a flow_group is eligible for submission wrt a given job.
 
         By default, an operation is eligible for submission when it
         is not considered active, that means already queued or running.
         """
-        if job_operation is None:
+        if flow_group is None or job is None:
             return False
-        if job_operation.get_status() >= JobStatus.submitted:
+        if flow_group._get_status(job) >= JobStatus.submitted:
             return False
+        group_ops = set(flow_group.operations.values())
+        for other_group in self._groups.values():
+            if group_ops & set(other_group.operations.values()):
```

Comment:
Aren't the groups already iterable over their operations? Because then we can express this much simpler.
```suggestion
            if group_ops & set(other_group):
```

Reply:
If not, then we should make them iterable over the operations.